### PR TITLE
[Frontend] Batch inference for llm.chat() API 

### LINKED ISSUE #8481: [Feature]: Batch inference for `llm.chat()` API
### 🚀 The feature, motivation and pitch

Currently `llm.chat()` API only supports one conversation per inference. This means we cannot use this API to fully leverage vLLM for efficient offline processing.

### Alternatives

_No response_

### Additional context

Implementation should be rather straightforward:

1. at API level, `llm.chat()` should also accept a list of conversations.
2. When `llm.chat()` is invoked, the list of conversations will be parsed into list of prompts, and all multimodal data items will be retrieved and loaded into their corresponding format that `llm.generate()` accepts.
3. Send the list of `{prompt: xxx, multi_modal_data: xxx}` to the `llm.generate()`

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by aandyw**
Currently `llm.chat()` API only supports one conversation per inference. This PR resolves this.

Changes:
- `llm.chat()` should also accept a list of conversations.
- When `llm.chat()` is invoked, the list of conversations will be parsed into list of prompts and loaded into their corresponding format that `llm.generate()` accepts.
- The return is a list of multiple output conversations

FIX #8481 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -277,15 +277,15 @@ def generate(
     def generate(
         self,
         prompts: Union[Union[PromptInputs, Sequence[PromptInputs]],
-                       Optional[Union[str, List[str]]]] = None,
+                       Optional[Union[str, List[str]]], ] = None,
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:280]: Please avoid formatting changes in sections of the code that are unrelated to this PR.

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -344,22 +344,23 @@ def generate(
...
+            guided_options=guided_options_request,
+        )
 
         outputs = self._run_engine(use_tqdm=use_tqdm)
         return LLMEngine.validate_outputs(outputs, RequestOutput)
 
     def chat(
         self,
-        messages: List[ChatCompletionMessageParam],
+        conversations: List[List[ChatCompletionMessageParam]],
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:355]: We should still allow single input, just like the case for `LLM.generate`.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for implementing this! Some comments.

File: vllm/entrypoints/llm.py (hunk 1 above)
Comment (aandyw) [reply vllm/entrypoints/llm.py:280]: I think `./format.sh` makes these changes? Should I not run the ruff linter + formatter

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:280]: `format.sh` should only run `ruff check` and format using `yapf`. It doesn't make those changes on my end.

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:280]: Make sure your IDE doesn't have auto-formatters enabled.

File: vllm/entrypoints/llm.py (hunk 2 above)
Comment (aandyw) [reply vllm/entrypoints/llm.py]: Does `Union[List[ChatCompletionMessageParam], List[List[ChatCompletionMessageParam]]]` make sense then?

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Yes.

Comment (aandyw) [reply vllm/entrypoints/llm.py]: Been getting this issue with my latest commit that I can't seem to resolve:
```
vllm/entrypoints/llm.py:399: error: Incompatible types in assignment (expression has type "Union[List[Union[Any, CustomChatCompletionMessageParam]], List[List[Union[Any, CustomChatCompletionMessageParam]]]]", variable has type "List[List[Union[Any, CustomChatCompletionMessageParam]]]")  [assignment]
vllm/entrypoints/llm.py:402: error: List item 0 has incompatible type "Union[List[Union[Any, CustomChatCompletionMessageParam]], List[List[Union[Any, CustomChatCompletionMessageParam]]]]"; expected "List[Union[Any, CustomChatCompletionMessageParam]]"  [list-item]
```

Any ideas? @DarkLight1337

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Your current `for`-loop is only designed to handled nested list, not single input.

File: vllm/entrypoints/llm.py [hunk 3]
```diff
@@ -386,49 +391,61 @@ def chat(
                 to each message.
 
         Returns:
-            A list of ``RequestOutput`` objects containing the generated
-            responses in the same order as the input messages.
+            A list of lists or single list of ``RequestOutput`` objects
+            containing the generated responses in the same order as the input
+            conversations and messages.
         """
+        list_of_conversations: List[
```
Comment (aandyw) [inline vllm/entrypoints/llm.py:398]: @DarkLight1337 I'm trying to address the single conversation case here but I'm getting the format issue:
```
vllm/entrypoints/llm.py:399: error: Incompatible types in assignment (expression has type "Union[List[Union[Any, CustomChatCompletionMessageParam]], List[List[Union[Any, CustomChatCompletionMessageParam]]]]", variable has type "List[List[Union[Any, CustomChatCompletionMessageParam]]]")  [assignment]
vllm/entrypoints/llm.py:402: error: List item 0 has incompatible type "Union[List[Union[Any, CustomChatCompletionMessageParam]], List[List[Union[Any, CustomChatCompletionMessageParam]]]]"; expected "List[Union[Any, CustomChatCompletionMessageParam]]"  [list-item]
```

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:398]: You tried to assign `list_of_conversations: List[List[...]] = conversations: Union[List[...], List[List[...]]]` before you normalized the type to `List[List[...]]`. Also, you should invert the `if` condition since we only want to create an additional layer of nesting if the inner item is *not* a list.

I suggest this instead:

```
list_of_conversations: List[List[ChatCompletionMessageParam]]

if is_list_of(conversations, list):
    # conversations is List[List[...]]
    list_of_conversations = conversations
else:
    # conversations is List[...]
    list_of_conversations = [conversations]
```

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -351,15 +351,16 @@ def generate(
...
+                             List[List[ChatCompletionMessageParam]]],
         sampling_params: Optional[Union[SamplingParams,
                                         List[SamplingParams]]] = None,
         use_tqdm: bool = True,
         lora_request: Optional[LoRARequest] = None,
         chat_template: Optional[str] = None,
         add_generation_prompt: bool = True,
         tools: Optional[List[Dict[str, Any]]] = None,
-    ) -> List[RequestOutput]:
+    ) -> Union[List[List[RequestOutput]], List[RequestOutput]]:
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:363]: Let's use overloads to better indicate the return type based on the type of `conversations`

**Comment by aandyw** [comment]
Thanks for the help @DarkLight1337 you're the goat 🐐 

Should be good now?

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -371,8 +372,9 @@ def chat(
         to the OpenAI API.
 
         Args:
-            messages: A single conversation represented as a list of messages.
-                Each message is a dictionary with 'role' and 'content' keys.
+            conversations: A list or a single conversation represented as a list
+                of messages. Each message is a dictionary with 'role' and
+                'content' keys.
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:377]:
```suggestion
            conversations: A list of conversations or a single conversation. 
                - Each conversation is represented as a list of messages.
                - Each message is a dictionary with 'role' and 'content' keys.
```

File: vllm/entrypoints/llm.py [hunk 6]
```diff
@@ -386,49 +388,66 @@ def chat(
                 to each message.
 
         Returns:
-            A list of ``RequestOutput`` objects containing the generated
-            responses in the same order as the input messages.
+            A list of lists or single list of ``RequestOutput`` objects
+            containing the generated responses in the same order as the input
+            conversations and messages.
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:393]: Actually, we should return a single list here. The `LLM.generate` method returns a single list regardless of single or multiple inputs.

Comment (aandyw) [reply vllm/entrypoints/llm.py:393]: so no overloads then?

Comment (aandyw) [reply vllm/entrypoints/llm.py:393]: Little confused here. Doesn't `LLM.generate()` return `List[RequestOutput]`? Wouldn't multiple inputs give a List of ` List[RequestOutput]` then?

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:393]: You can see that `LLM.generate` returns `List[RequestOutput]` even when multiple `prompts` are given.

Comment (aandyw) [reply vllm/entrypoints/llm.py:393]: Are you saying we should return `List[List[RequestOutput]]` here? Because both single and multiple items will give this result.

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:393]: Oh, I see what you mean now, sorry for the confusion. In that case then yes, it should return a single list for single input. Can you add overloads as mentioned above?

**Comment by DarkLight1337** [comment]
Thanks for the effort. Finally, can you add tests to check the behavior of batch inference for `LLM.chat`?

File: vllm/entrypoints/llm.py [hunk 7]
```diff
@@ -349,9 +349,10 @@ def generate(
         outputs = self._run_engine(use_tqdm=use_tqdm)
         return LLMEngine.validate_outputs(outputs, RequestOutput)
 
+    @overload
     def chat(
         self,
-        messages: List[ChatCompletionMessageParam],
+        conversations: List[ChatCompletionMessageParam],
```
Comment (mgoin) [inline vllm/entrypoints/llm.py:355]: We should note this as a (small) breaking change in the release for people who might be using kwargs for messages

Comment (ywang96) [reply vllm/entrypoints/llm.py:355]: @aandyw I agree with Michael with this. Why did we need to rename this to conversations in the first place? I think messages fit better since a lot of people are used to the OpenAI format (`prompt` for completions, `messages` for chat completions.

Comment (aandyw) [reply vllm/entrypoints/llm.py:355]: Makes sense. I'll change it back. I thought it would be more clear that it's possible to hold multiple separate conversations/messages

**Comment by DarkLight1337** [comment]
Please run `format.sh` to fix the lint errors.

**Review by mgoin — APPROVED** [review]
This looks good to me, thanks for reverting to messages

File: vllm/entrypoints/llm.py [hunk 8]
```diff
@@ -389,49 +419,66 @@ def chat(
...
-                tokenizer,
-                conversation=conversation,
-                chat_template=chat_template,
-                add_generation_prompt=add_generation_prompt,
-                tools=tools,
-            )
+            # messages is List[...]
+            list_of_messages = [messages]
+
+        outputs = []
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:436]:
```suggestion
        outputs: List[List[RequestOutput]]] = []
```

File: vllm/entrypoints/llm.py [hunk 9]
```diff
@@ -389,49 +419,66 @@ def chat(
...
-                tokenizer,
-                conversation=conversation,
-                chat_template=chat_template,
-                add_generation_prompt=add_generation_prompt,
-                tools=tools,
-            )
+            # messages is List[...]
+            list_of_messages = [messages]
+
+        outputs: List[List[RequestOutput]]] = []
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:436]:
```suggestion
        outputs: List[List[RequestOutput]] = []
```

**Comment by ywang96** [comment]
@aandyw Hey thanks for the contribution! 

Can you make a follow-up PR to update the example for this feature in https://github.com/vllm-project/vllm/blob/main/examples/offline_inference_chat.py?

**Comment by DarkLight1337** [comment]
Actually can you merge from main to fix the CI failure and add the example here?

**Comment by ywang96** [comment]
@DarkLight1337 I have updated this PR with main and example

File: vllm/entrypoints/llm.py [hunk 10]
```diff
@@ -391,49 +421,66 @@ def chat(
...
 
-        if mm_data is not None:
-            prompt["multi_modal_data"] = mm_data
+            out = self.generate(
+                prompt,
+                sampling_params=sampling_params,
+                use_tqdm=use_tqdm,
+                lora_request=lora_request,
+            )
+            outputs.append(out)
```
Comment (ywang96) [inline vllm/entrypoints/llm.py:480]: Hey @aandyw! I realized after some testing that this implementation is actually not what we want.

Here we're still parsing the list of messages, and calling `.generate` one by one, but what we really want is to make a batch of requests after parsing the list of messages, and call `.generate` on that one batch all together.

Can you make the changes accordingly? Thanks!

**Review by ywang96 — CHANGES_REQUESTED** [review]

File: vllm/entrypoints/llm.py (hunk 10 above)
Comment (aandyw) [reply vllm/entrypoints/llm.py:480]: Sounds good. The overall output from `llm.chat` should be the same correct?

Comment (ywang96) [reply vllm/entrypoints/llm.py:480]: Yep!

**Comment by DarkLight1337** [comment]
Do you still want to return a single list when a single conversation is passed?

**Comment by ywang96** [comment]
> Do you still want to return a single list when a single conversation is passed?

I discussed with @aandyw offline about it. We agreed that `.chat` should just return the same output format `.generate` does.

File: vllm/entrypoints/llm.py [hunk 11]
```diff
@@ -523,47 +525,63 @@ def chat(
...
-        return self.generate(
-            inputs,
+        outputs = self.generate(
+            prompts,
             sampling_params=sampling_params,
             use_tqdm=use_tqdm,
             lora_request=lora_request,
         )
 
+        return outputs
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:583]:
```suggestion
        return self.generate(
            prompts,
            sampling_params=sampling_params,
            use_tqdm=use_tqdm,
            lora_request=lora_request,
        )
```

**Review by ywang96 — APPROVED** [review]

**Comment by ywang96** [comment]
@patrickvonplaten FYI - with this PR merged, users will be able to run efficient batch processing with Pixtral!

**Review by mgoin — APPROVED** [review]

**Comment by aandyw** [comment]
@ywang96 @DarkLight1337 @mgoin Thank you guys for all the help on this PR 🙏

**Comment by patrickvonplaten** [comment]
Great addition!
